jit/gen.c

```diff
@@ -449,35 +451,55 @@ static inline bool gen_vec(enum arg rm, enum arg reg, void (*helper)(), gadget_t
         gadget = (*helper_gadgets_mem)[v_reg];
     } else {
         extern gadget_t vec_helper_reg_gadgets[vec_arg_count][vec_arg_count];
-        gadget = vec_helper_reg_gadgets[v_rm][v_reg];
+        gadget = vec_helper_reg_gadgets[v_reg][v_rm];
     }
-    if (gadget == NULL)
+    if (gadget == NULL) {
         UNDEFINED;
+    }
 
-    if (v_rm != vec_arg_mem) {
-        GEN(gadget);
-        GEN(helper);
-        GEN((modrm->opcode * sizeof(union xmm_reg)) |
-                ((modrm->rm_opcode * sizeof(union xmm_reg) << 8)));
-    } else {
-        gen_addr(state, modrm, seg_gs, saved_ip);
-        GEN(gadget);
-        GEN(saved_ip);
-        GEN(helper);
-        GEN(modrm->opcode * sizeof(union xmm_reg));
+    switch (v_rm) {
+        case vec_arg_xmm:
+            GEN(gadget);
+            GEN(helper);
+            GEN((modrm->opcode * sizeof(union xmm_reg))
+                    | (modrm->rm_opcode * sizeof(union xmm_reg) << 8));
+            break;
+
+        case vec_arg_mem:
+            gen_addr(state, modrm, seg_gs, saved_ip);
+            GEN(gadget);
+            GEN(saved_ip);
+            GEN(helper);
+            GEN(modrm->opcode * sizeof(union xmm_reg));
+            break;
+
+        case vec_arg_imm:
+            // TODO: support immediates and opcode
+            GEN(gadget);
+            GEN(helper);
+            GEN((modrm->rm_opcode * sizeof(union xmm_reg))
+                    | (((uint16_t) imm) << 8));
+            break;
+
+        default: die("unimplemented vecarg");
     }
     return true;
 }
 
 #define _v(src, dst, helper, helper_gadgets, z) do { \
     extern gadget_t helper_gadgets[vec_arg_count]; \
-    if (!gen_vec(src, dst, helper, &helper_gadgets, state, &modrm, saved_ip, seg_gs)) return false; \
+    if (!gen_vec(src, dst, helper, &helper_gadgets, state, &modrm, 0, saved_ip, seg_gs)) return false; \
+} while (0)
+#define _vi(imm, dst, helper, helper_gadgets, z) do { \
```

Comment:
can you make this _v_imm too. pretty sure that's the last nit

Reply:
done